packages/zowe-explorer-api/src/utils/ErrorCorrelator.ts

```diff
@@ -0,0 +1,360 @@
+     */
+    resources?: ExternalResource[];
+}
+
+export interface CorrelatedErrorProps {
+    errorCode?: string;
+    correlation?: Omit<ErrorCorrelation, "matches">;
+    initialError: Error | string;
+}
+
+export interface CorrelateErrorOpts {
+    profileType?: string;
+    templateArgs?: Record<string, string>;
+}
+
+export interface DisplayErrorOpts extends CorrelateErrorOpts {
+    additionalContext?: string;
+    allowRetry?: boolean;
+}
+
+export interface DisplayCorrelatedErrorOpts extends Omit<DisplayErrorOpts, "profileType"> {}
+
+export interface HandledErrorInfo {
+    correlation: CorrelatedError;
+    userResponse: string | undefined;
+}
+
+/**
+ * Representation of the given error as a correlated error (wrapper around the `Error` class).
+ *
+ * Used to cache the error info such as tips, the match that was encountered and the full error message.
+ */
+export class CorrelatedError {
+    public errorCode?: string;
+    public message: string;
+    private wasCorrelated: boolean;
+
+    public constructor(public properties: CorrelatedErrorProps) {
+        this.errorCode = properties.initialError instanceof ImperativeError ? properties.initialError.errorCode : this.properties.errorCode;
+        this.wasCorrelated = properties.correlation != null;
+
+        if (this.wasCorrelated) {
+            this.message = this.properties.correlation.summary;
+        } else {
+            this.message = this.properties.initialError instanceof Error ? this.properties.initialError.message : this.properties.initialError;
+        }
+    }
+
+    public get correlationFound(): boolean {
+        return this.wasCorrelated;
+    }
+
+    public get stack(): string | undefined {
+        return this.initial instanceof Error ? this.initial.stack : undefined;
+    }
+
+    public get initial(): Error | string {
+        return this.properties.initialError;
+    }
+
+    public asError(): Error {
+        const err = new Error(this.message);
+        err.stack = this.stack;
+        return err;
+    }
+
+    public toString(): string {
+        return this.message;
+    }
+}
+
+export enum ZoweExplorerApiType {
+    Mvs = "mvs",
+    Jes = "jes",
+    Uss = "uss",
+    Command = "cmd",
+    /* errors that match all API types */
+    All = "all",
+}
+
+export type ErrorsForApiType = Map<ZoweExplorerApiType, ApiErrors>;
+export type ApiErrors = Record<string, ErrorCorrelation[]>;
+
+export class ErrorCorrelator {
+    private static instance: ErrorCorrelator = null;
+
+    private errorMatches: ErrorsForApiType = new Map([
+        [
+            ZoweExplorerApiType.Mvs,
+            {
+                zosmf: [
+                    {
+                        errorCode: "500",
+                        matches: ["Client is not authorized for file access.", /An I\/O abend was trapped\.(.+?)\n(.+?)__code=0x0913/],
+                        summary: "Insufficient write permissions for this data set. The data set may be read-only or locked.",
```

Comment:
This is a pain point I ran into when writing these error correlations. Since we don't have l10n set up for ZE API, we can't localize these strings. It'd be nice to add localization to the API, but since this would be a large, additional effort, I'll track this as an enhancement 🙂

Reply:
Thanks for considering this 🙏

Reply:
Created:
-  https://github.com/zowe/zowe-explorer-vscode/pull/3243#:~:text=ZE%2DAPI%3A%20Allow%20strings%20in%20the%20API%20to%20be%20localized%20%233283